File: app/src/main/java/com/brainwellnessspa/ResourceModule/Fragments/AppsFragment.java
package com.brainwellnessspa.ResourceModule.Fragments;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.SystemClock;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.brainwellnessspa.BWSApplication;
import com.brainwellnessspa.R;
import com.brainwellnessspa.ResourceModule.Activities.ResourceDetailsActivity;
import com.brainwellnessspa.ResourceModule.Models.ResourceListModel;
import com.brainwellnessspa.Utility.APIClient;
import com.brainwellnessspa.Utility.CONSTANTS;
import com.brainwellnessspa.Utility.MeasureRatio;
import com.brainwellnessspa.databinding.AppsListLayoutBinding;
import com.brainwellnessspa.databinding.FragmentAppsBinding;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AppsFragment extends Fragment {
    FragmentAppsBinding binding;
    String apps, UserID, Category;
    private long mLastClickTime = 0;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_apps, container, false);
        View view = binding.getRoot();
        Bundle bundle = this.getArguments();
        if (bundle != null) {
            apps = bundle.getString("apps");
            UserID = bundle.getString("UserID");
            Category = bundle.getString("Category");
        }
        GridLayoutManager manager = new GridLayoutManager(getActivity(), 2);
        binding.rvAppsList.setLayoutManager(manager);
        binding.rvAppsList.setItemAnimator(new DefaultItemAnimator());
        prepareData();
        return view;
    }

    void prepareData() {
        BWSApplication.showProgressBar(binding.progressBar, binding.progressBarHolder, getActivity());
        Call<ResourceListModel> listCall = APIClient.getClient().getResourcLists(UserID, CONSTANTS.FLAG_FIVE, Category);
        listCall.enqueue(new Callback<ResourceListModel>() {
            @Override
            public void onResponse(Call<ResourceListModel> call, Response<ResourceListModel> response) {
                if (response.isSuccessful()) {
                    BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, getActivity());
                    ResourceListModel listModel = response.body();
                    AppsAdapter adapter = new AppsAdapter(listModel.getResponseData(), getActivity(), apps);
                    binding.rvAppsList.setAdapter(adapter);

                    if (listModel.getResponseData().size() != 0) {
                        binding.llError.setVisibility(View.GONE);
                        binding.rvAppsList.setVisibility(View.VISIBLE);
                    } else {
                        binding.llError.setVisibility(View.VISIBLE);
                        binding.rvAppsList.setVisibility(View.GONE);
                    }
                }
            }

            @Override
            public void onFailure(Call<ResourceListModel> call, Throwable t) {
                BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, getActivity());
            }
        });
    }

    public class AppsAdapter extends RecyclerView.Adapter<AppsAdapter.MyViewHolder> {
        List<ResourceListModel.ResponseData> listModelList;
        Context ctx;
        String apps;

        public AppsAdapter(List<ResourceListModel.ResponseData> listModelList, Context ctx, String apps) {
            this.listModelList = listModelList;
            this.ctx = ctx;
            this.apps = apps;
        }

        @NonNull
        @Override
        public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            AppsListLayoutBinding v = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext())
                    , R.layout.apps_list_layout, parent, false);
            return new MyViewHolder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
            holder.binding.tvTitle.setText(listModelList.get(position).getTitle());
            MeasureRatio measureRatio = BWSApplication.measureRatio(ctx, 0,
                    1, 1, 0.44f, 0);
            holder.binding.ivRestaurantImage.getLayoutParams().height = (int) (measureRatio.getHeight() * measureRatio.getRatio());
            holder.binding.ivRestaurantImage.getLayoutParams().width = (int) (measureRatio.getWidthImg() * measureRatio.getRatio());
            holder.binding.ivRestaurantImage.setScaleType(ImageView.ScaleType.FIT_XY);

            Glide.with(ctx).load(listModelList.get(position).getImage()).thumbnail(0.05f)
                    .diskCacheStrategy(DiskCacheStrategy.ALL).skipMemoryCache(false).into(holder.binding.ivRestaurantImage);
            holder.binding.rlMainLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (SystemClock.elapsedRealtime() - mLastClickTime < 1000) {
                        return;
                    }
                    mLastClickTime = SystemClock.elapsedRealtime();
                    Intent i = new Intent(getActivity(), ResourceDetailsActivity.class);
                    i.putExtra("apps", apps);
                    i.putExtra("title", listModelList.get(position).getTitle());
                    i.putExtra("linkOne", listModelList.get(position).getResourceLink1());
                    i.putExtra("linkTwo", listModelList.get(position).getResourceLink2());
                    i.putExtra("image", listModelList.get(position).getImage());
                    i.putExtra("description", listModelList.get(position).getDescription());
                    startActivity(i);
                }
            });
        }

        @Override
        public int getItemCount() {
            return listModelList.size();
        }

        public class MyViewHolder extends RecyclerView.ViewHolder {
            AppsListLayoutBinding binding;

            public MyViewHolder(AppsListLayoutBinding binding) {
                super(binding.getRoot());
                this.binding = binding;
            }
        }
    }
}

File: app/src/main/java/com/brainwellnessspa/resourceModule/Fragments/AppsFragment.kt
package com.brainwellnessspa.resourceModule.Fragments

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.os.SystemClock
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.brainwellnessspa.BWSApplication
import com.brainwellnessspa.R
import com.brainwellnessspa.resourceModule.activities.ResourceDetailsActivity
import com.brainwellnessspa.resourceModule.models.ResourceListModel
import com.brainwellnessspa.utility.APINewClient
import com.brainwellnessspa.utility.CONSTANTS
import com.brainwellnessspa.databinding.AppsListLayoutBinding
import com.brainwellnessspa.databinding.FragmentAppsBinding
import com.bumptech.glide.Glide
import com.bumptech.glide.Priority
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestOptions
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class AppsFragment : Fragment() {
    lateinit var binding: FragmentAppsBinding
    var apps: String? = ""
    var USERID: String? = ""
    var CoUserID: String? = ""
    var Category: String? = ""
    private var mLastClickTime: Long = 0
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_apps, container, false)
        val view = binding.root
        val bundle = this.arguments
        if (bundle != null) {
            apps = bundle.getString("apps")
            Category = bundle.getString("Category")
        }
        val shared1: SharedPreferences =
            requireActivity().getSharedPreferences(CONSTANTS.PREFE_ACCESS_SIGNIN_COUSER, Context.MODE_PRIVATE)
        USERID = shared1.getString(CONSTANTS.PREFE_ACCESS_mainAccountID, "")
        CoUserID = shared1.getString(CONSTANTS.PREFE_ACCESS_UserId, "")
        val manager = GridLayoutManager(activity, 2)
        binding.rvAppsList.layoutManager = manager
        binding.rvAppsList.itemAnimator = DefaultItemAnimator()
        return view
    }

    override fun onResume() {
        prepareData()
        super.onResume()
    }

    fun prepareData() {
        BWSApplication.showProgressBar(binding.progressBar, binding.progressBarHolder, activity)
        val listCall =
            APINewClient.getClient().getResourceList(CoUserID, CONSTANTS.FLAG_FIVE, Category)
        listCall.enqueue(object : Callback<ResourceListModel?> {
            override fun onResponse(
                call: Call<ResourceListModel?>,
                response: Response<ResourceListModel?>
            ) {
                try {
                    val listModel = response.body()
                    if (listModel!!.responseCode.equals(
                            getString(R.string.ResponseCodesuccess),
                            ignoreCase = true
                        )
                    ) {
                        BWSApplication.hideProgressBar(
                            binding.progressBar,
                            binding.progressBarHolder,
                            activity
                        )
                        val adapter = AppsAdapter(listModel.responseData, activity, apps)
                        binding.rvAppsList.adapter = adapter
                        if (listModel.responseData!!.isNotEmpty()) {
                            binding.llError.visibility = View.GONE
                            binding.rvAppsList.visibility = View.VISIBLE
                        } else {
                            binding.llError.visibility = View.VISIBLE
                            binding.rvAppsList.visibility = View.GONE
                        }
                    } else if (listModel.responseCode.equals(
                            getString(R.string.ResponseCodefail),
                            ignoreCase = true
                        )
                    ) {
                        BWSApplication.hideProgressBar(
                            binding.progressBar,
                            binding.progressBarHolder,
                            activity
                        )
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

            override fun onFailure(call: Call<ResourceListModel?>, t: Throwable) {
                BWSApplication.hideProgressBar(
                    binding.progressBar,
                    binding.progressBarHolder,
                    activity
                )
            }
        })
    }

    inner class AppsAdapter(
        var listModelList: List<ResourceListModel.ResponseData>?,
        var ctx: Context?,
        var apps: String?
    ) : RecyclerView.Adapter<AppsAdapter.MyViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
            val v: AppsListLayoutBinding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.apps_list_layout,
                parent,
                false
            )
            return MyViewHolder(v)
        }

        override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
            holder.binding.tvTitle.text = listModelList!![position].title
            val measureRatio = BWSApplication.measureRatio(ctx, 0f, 1f, 1f, 0.42f, 0f)
            holder.binding.ivRestaurantImage.layoutParams.height =
                (measureRatio.height * measureRatio.ratio).toInt()
            holder.binding.ivRestaurantImage.layoutParams.width =
                (measureRatio.widthImg * measureRatio.ratio).toInt()
            holder.binding.ivRestaurantImage.scaleType = ImageView.ScaleType.FIT_XY
            Glide.with(ctx!!).load(listModelList!![position].image).thumbnail(0.05f)
                .apply(RequestOptions.bitmapTransform(RoundedCorners(40))).priority(Priority.HIGH)
                .diskCacheStrategy(DiskCacheStrategy.ALL).skipMemoryCache(false)
                .into(holder.binding.ivRestaurantImage)
            holder.binding.rlMainLayout.setOnClickListener(View.OnClickListener {
                if (SystemClock.elapsedRealtime() - mLastClickTime < 1000) {
                    return@OnClickListener
                }
                mLastClickTime = SystemClock.elapsedRealtime()
                val i = Intent(activity, ResourceDetailsActivity::class.java)
                i.putExtra("apps", apps)
                i.putExtra("id", listModelList!![position].iD)
                i.putExtra("title", listModelList!![position].title)
                i.putExtra("linkOne", listModelList!![position].resourceLink1)
                i.putExtra("linkTwo", listModelList!![position].resourceLink2)
                i.putExtra("image", listModelList!![position].detailimage)
                i.putExtra("description", listModelList!![position].description)
                i.putExtra("mastercat", listModelList!![position].masterCategory)
                i.putExtra("subcat", listModelList!![position].subCategory)
                startActivity(i)
                activity!!.overridePendingTransition(0, 0)
            })
        }

        override fun getItemCount(): Int {
            return listModelList!!.size
        }

        inner class MyViewHolder(var binding: AppsListLayoutBinding) :
            RecyclerView.ViewHolder(binding.root)
    }
}